src/common/registry/test/validator-registry/registry-update.spec.ts

```diff
@@ -339,3 +324,90 @@ describe('Empty registry', () => {
     await registryService.update(address, blockHash);
   });
 });
+
+describe('Reorg detection', () => {
+  const provider = new JsonRpcBatchProvider(process.env.PROVIDERS_URLS);
+  let registryService: ValidatorRegistryService;
+  let registryStorageService: RegistryStorageService;
+  let moduleRef: TestingModule;
+  const mockCall = jest.spyOn(provider, 'call').mockImplementation(async () => '');
+  const CHAIN_ID = process.env.CHAIN_ID || 1;
+  const address = REGISTRY_CONTRACT_ADDRESSES[CHAIN_ID];
+  let mikroOrm: MikroORM;
+
+  jest.spyOn(provider, 'detectNetwork').mockImplementation(async () => getNetwork('mainnet'));
+
+  beforeEach(async () => {
+    const imports = [
+      MikroOrmModule.forRoot(mikroORMConfig),
+      MockLoggerModule.forRoot({
+        log: jest.fn(),
+        error: jest.fn(),
+        warn: jest.fn(),
+      }),
+      ValidatorRegistryModule.forFeature({ provider }),
+    ];
+    moduleRef = await Test.createTestingModule({
+      imports,
+      providers: [{ provide: LOGGER_PROVIDER, useValue: {} }],
+    }).compile();
+    registryService = moduleRef.get(ValidatorRegistryService);
+    registryStorageService = moduleRef.get(RegistryStorageService);
+    mikroOrm = moduleRef.get(MikroORM);
+    const generator = mikroOrm.getSchemaGenerator();
+    await generator.updateSchema();
+  });
+
+  afterEach(async () => {
+    mockCall.mockReset();
+    await clearDb(mikroOrm);
+    await registryStorageService.onModuleDestroy();
+  });
+
+  test('init on update', async () => {
+    const saveRegistryMock = jest.spyOn(registryService, 'saveOperators');
+    const saveKeyRegistryMock = jest.spyOn(registryService, 'saveKeys');
+    const finalizedUsedSigningKeys = 1;
+
+    const keysWithModuleAddress = keys.map((key) => {
+      return { ...key, moduleAddress: address };
+    });
+
+    const operatorsWithModuleAddress = operators.map((key) => {
+      return { ...key, moduleAddress: address, finalizedUsedSigningKeys };
+    });
+
+    const unrefMock = registryServiceMock(moduleRef, provider, {
+      keys: keysWithModuleAddress,
+      operators: operatorsWithModuleAddress,
+    });
+
+    await registryService.update(address, blockHash);
+
+    expect(saveRegistryMock).toBeCalledTimes(1);
+    expect(saveKeyRegistryMock.mock.calls.length).toBeGreaterThanOrEqual(1);
+
+    await compareTestKeysAndOperators(address, registryService, {
+      keys: keysWithModuleAddress,
+      operators: operatorsWithModuleAddress,
+    });
+
+    unrefMock();
+
+    const keysWithSpoiledLeftEdge = clone(keysWithModuleAddress).map((key) =>
+      key.index >= finalizedUsedSigningKeys ? { ...key } : { ...key, key: '', depositSignature: '' },
+    );
+
+    registryServiceMock(moduleRef, provider, {
+      keys: keysWithSpoiledLeftEdge,
+      operators: operatorsWithModuleAddress,
+    });
+
+    await registryService.update(address, blockHash);
```

Comment:
keysWithSpoiledLeftEdge - keysModifiedBeforeFinalized or smth like this will be more clear (keysModifiedBeforeFinalizedUsedSigningKeys  ). it took for we a lot of time to understand that was spoiled

Reply:
here I am purposely "breaking" the data to see that the  `update` will not take this data.

Reply:
added a comment to the tests to make it clearer.